File: backend/src/utils/seededRandom.js
/**
 * seededRandom.js — Deterministic PRNG for simulated executor outcomes.
 *
 * Uses mulberry32 algorithm seeded by hashing case._id.
 * Same case always resolves to the same simulated outcome on every run,
 * so the batch recovery figure is stable and reproducible.
 */

/**
 * Simple string hash → 32-bit integer.
 * Produces a consistent numeric seed from any string (like a MongoDB ObjectId).
 */
function hashString(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash; // Convert to 32-bit integer
  }
  return Math.abs(hash);
}

/**
 * mulberry32 — a fast, deterministic 32-bit PRNG.
 * Returns a function that produces a new pseudo-random float [0, 1) on each call.
 *
 * @param {number} seed - A 32-bit integer seed.
 * @returns {function} A function that returns the next pseudo-random number [0, 1).
 */
function mulberry32(seed) {
  return function () {
    seed |= 0;
    seed = (seed + 0x6D2B79F5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/**
 * Create a seeded random number generator from a case ID string.
 *
 * @param {string} caseId - The case's _id as a string (e.g., MongoDB ObjectId.toString()).
 * @returns {function} A function that returns deterministic pseudo-random floats [0, 1).
 *
 * Usage:
 *   const rng = createSeededRng(case._id.toString());
 *   const outcome = rng(); // always the same for this case ID
 *   const isPaid = outcome < 0.65; // ~65% chance of "paid"
 */
function createSeededRng(caseId) {
  const seed = hashString(caseId);
  return mulberry32(seed);
}

module.exports = { createSeededRng, hashString, mulberry32 };

File: backend/app/utils/seeded_random.py
import math

def hash_string(s: str) -> int:
    """Simple string hash -> 32-bit integer."""
    h = 0
    for char in s:
        h = ((h << 5) - h) + ord(char)
        h = h & 0xFFFFFFFF # 32-bit uint mask
    # Convert uint32 to int32 to match JS Math.abs bitwise behavior more closely, 
    # but since JS does `abs`, we just do abs on the signed 32bit int.
    # Actually in JS `hash & hash` converts to signed 32-bit. 
    signed_h = h if h < 0x80000000 else h - 0x100000000
    return abs(signed_h)

def create_seeded_rng(case_id: str):
    """
    mulberry32 - a fast, deterministic 32-bit PRNG.
    Returns a function that produces a new pseudo-random float [0, 1) on each call.
    """
    seed = hash_string(case_id)
    
    def rng():
        nonlocal seed
        seed = (seed + 0x6D2B79F5) & 0xFFFFFFFF
        
        # We need JS imul behavior: signed 32-bit multiplication
            
        def signed_32(val):
            val = val & 0xFFFFFFFF
            return val if val < 0x80000000 else val - 0x100000000
            
        def unsigned_right_shift(val, shift):
            return (val & 0xFFFFFFFF) >> shift

        def python_imul(a, b):
            return signed_32(signed_32(a) * signed_32(b))

        t = python_imul(seed ^ unsigned_right_shift(seed, 15), 1 | seed)
        t = (t + python_imul(t ^ unsigned_right_shift(t, 7), 61 | t)) ^ t
        
        result = ((t ^ unsigned_right_shift(t, 14)) & 0xFFFFFFFF) / 4294967296.0
        return result

    return rng
